Debug node counter

shayanhabibi opened the pull request:
I've added a debug node counter; it demonstrates a memory leak of the nodes.

I'll investigate the leak and absolve it when found.

shayanhabibi committed:
Debug counter added; leak reduced

added node counter with debug procs
Corrected double inc; still leaking
Fixed leak from 29 to 2+/-2

disruptek commented on loony.nim:
```diff
@@ -273,15 +275,15 @@ proc popImpl[T](queue: LoonyQueue[T]; forcedCoherance: static bool = false): T =
         break
 
-proc pop*[T](queue: LoonyQueue[T]): T =
+template pop*[T](queue: LoonyQueue[T]): T =
```
These should really just stay procedures for a few reasons:
- compilers can optimize them as what they are -- procedures
- no code bloat from invoking them all over your program
- symbols in traceback/profiling become more useful
- capturing local symbols so we don't have to play games with imports

The cost of invoking the procedure is not significant and if a time comes when there is a performance advantage to running or optimizing only part of the procedure, sure, we can export multiple templates in addition to the existing api.

disruptek commented on loony/node.nim:
```diff
@@ -5,10 +5,27 @@ import loony/memalloc
     next*  : Atomic[NodePtr]            # NodePtr - successor node
     ctrl*  : ControlBlock               # Control block for mem recl
 
+when defined(loonyDebug):
+  var nodeCounter* {.global.}: Atomic[int]
+  nodeCounter.store(0)
+
+  proc echoDebugNodeCounter*(expected: int = 0) =
+    ## This will output the counter if it is above the expected value
+    let counter = nodeCounter.load()
+    if counter > 0:
+      warn "Node counter: " & $nodeCounter.load()
+  
+  template debugNodeCounter*(body: untyped) =
+    let initC = nodeCounter.load()
+    body
+    let newC = nodeCounter.load()
+    if (newC - initC) > 0:
+      warn "Finished block with node count of " & $(nodeCounter.load() - initC)
+
```
if you add
```nim
else:
  template debugNodeCounter(body: untyped): untyped = discard
```
...or whatever, then you can use it without the gratuitous `when defined(loonyDebug):`, right?

disruptek commented on tests/test.nim:
```diff
@@ -48,8 +48,8 @@ var counter {.global.}: Atomic[int]
-      when defined(danger):
-        1
+      when defined(danger) and false: # I'm the only person using windows
+        1                             # and this is annoying
```
If we don't need to sleep, feel free to take this stuff out...  I did it this way because we cannot sleep for less than 1ms on Windows. :shrug:

shayanhabibi committed:
Fixed leak; added documentation for advTail

Added helpers to determine leak source
incDeqCounter should fetchHead
Reimplemented advTail; leak sealed
Documentation added

shayanhabibi commented:
Very odd; testing the last commit with 10_000 continuations on my MacBook Pro 2016 gives the same memory leak I was having originally as opposed to when I was testing on my Ryzen 5 Windows 10 which was giving leaks in the range of 2+/-2

shayanhabibi commented:
Fixed the leak; tested on Windows and Mac Os.

Documentation and styling to follow. Pull can be completed whenever though since this is a pretty big bug.

shayanhabibi committed:
updated readme and tag
